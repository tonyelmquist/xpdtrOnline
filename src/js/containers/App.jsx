import React from "react";
import ReactDOM from "react-dom";
import { connect } from 'react-redux';
import {
  Button,
  Segment,
  Container,
  Grid,
} from "semantic-ui-react";
import PropTypes from 'prop-types';
import axios from "axios";
import SelectedList from "../components/SelectedList";
import {setCurrencies, selectCurrency, deselectCurrency, deselectAll} from '../actions/currencies';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: true
    };
  }

  componentDidMount() {
    axios
      .get(`https://data.norges-bank.no/api/data/EXR?lastNObservations=1`) // fetch the current XKCD comic. The site does not support CORS requests, so we make the request via a pass-through node server
      .then(response => {
        this.props.dispatch(setCurrencies(response.data));
      })
      .catch(function(error) {
        console.log(error);
      });
  }

  selectCurrency = currency => {
    this.props.dispatch(selectCurrency(currency));
  };

  deselectCurrency = currency => {
    this.props.dispatch(deselectCurrency(currency));
  };

  deselectAll = () => {
    this.props.dispatch(deselectAll());
  };

  render() {
    return (
      <Container>
        <Grid>
          <Grid.Row>
            <Grid.Column width={1} />
            <Grid.Column width={6}>
              <h1>Currencies</h1>
              <p>click to select</p>
            </Grid.Column>
            <Grid.Column width={2} />
            <Grid.Column width={6}>
              <h1>Selected Currencies</h1>
              <span>click to de-select</span>
              <Button
                className="deselect-button"
                onClick={() => this.deselectAll()}
                color="blue"
              >
                deselect all
              </Button>
            </Grid.Column>
            <Grid.Column width={1} />
          </Grid.Row>
          <Grid.Row>
            <Grid.Column width={1} />
            <Grid.Column width={6}>
             <SelectedList
                currencies={this.props.currencies}
                handleClick={this.selectCurrency}
                returnSelected={false}
              />
            </Grid.Column>
            <Grid.Column width={2} />
            <Grid.Column width={6}>
              <SelectedList
                currencies={this.props.currencies}
                handleClick={this.deselectCurrency}
                returnSelected={true}
              />
            </Grid.Column>
            <Grid.Column width={1} />
          </Grid.Row>
        </Grid>
      </Container>
    );
  }
}

App.propTypes = {
    dispatch: PropTypes.func.isRequired,
    currencies: PropTypes.array.isRequired
};

const mapStateToProps = state => ({
    currencies: state.currencies,
});

export default connect(mapStateToProps)(App);
