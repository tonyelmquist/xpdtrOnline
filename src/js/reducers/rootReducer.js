import Actions from "../constants/ReduxConstants";
import convert from "xml-js";

const defaultState = {
  currencies: []
};

export default function currencies(state = defaultState, action) {
  let newCurrencies;
  switch (action.type) {
    case Actions.SELECT_CURRENCY:
      newCurrencies = state.currencies.map(currencyItem => {
        return currencyItem.baseCurrency === action.currencyName
          ? {
              baseCurrency: currencyItem.baseCurrency,
              quoteCurrency: currencyItem.quoteCurrency,
              timePeriod: currencyItem.timePeriod,
              rate: currencyItem.rate,
              selected: true
            }
          : currencyItem;
      });
      return {
        currencies: newCurrencies
      };
    case Actions.DESELECT_CURRENCY:
      newCurrencies = state.currencies.map(currencyItem => {
        return currencyItem.baseCurrency === action.currencyName
          ? {
              baseCurrency: currencyItem.baseCurrency,
              quoteCurrency: currencyItem.quoteCurrency,
              timePeriod: currencyItem.timePeriod,
              rate: currencyItem.rate,
              selected: false
            }
          : currencyItem;
      });
      return {
        currencies: newCurrencies
      };
    case Actions.SET_CURRENCIES:
      const result = convert.xml2js(action.currencyData, { // converting XML to JSON rather than retrieving JSON directly because the JSON from the API was difficult to parse
        compact: true,
        spaces: 4
      });
      const dataSet =
        result["message:StructureSpecificData"]["message:DataSet"]["Series"]; //this is where the series are located in the JSON once the XML has been parsed
      let parsedData = dataSet.map(item => {
        const quoteItem = {
          baseCurrency: item._attributes.BASE_CUR,
          quoteCurrency: item._attributes.QUOTE_CUR,
          timePeriod: item.Obs._attributes.TIME_PERIOD,
          rate: item.Obs._attributes.OBS_VALUE,
          selected: false
        };
        return quoteItem;
      });
      const filterTime = parsedData[parsedData.length - 1].timePeriod;
      parsedData = parsedData.filter(item => item.timePeriod === filterTime);
      return {
        currencies: parsedData
      };
    case Actions.DESELECT_ALL:
      newCurrencies = state.currencies.map(currencyItem => {
        return {
          baseCurrency: currencyItem.baseCurrency,
          quoteCurrency: currencyItem.quoteCurrency,
          timePeriod: currencyItem.timePeriod,
          rate: currencyItem.rate,
          selected: false
        };
      });
      return {
        currencies: newCurrencies
      };
    default:
      return state;
  }
}
