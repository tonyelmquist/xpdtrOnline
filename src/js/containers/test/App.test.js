import React from "react";
import renderer from 'react-test-renderer';
import {createStore} from 'redux';
import App from '../App';
import rootReducer from '../../reducers/rootReducer';

const store = createStore(rootReducer);

it('should render correctly', () => {
  const component = renderer.create(<App store={store} />)
  expect(component.toJSON()).toMatchSnapshot()
})