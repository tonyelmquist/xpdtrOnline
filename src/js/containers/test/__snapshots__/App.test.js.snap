// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="ui container"
>
  <div
    className="ui grid"
  >
    <div
      className="row"
    >
      <div
        className="one wide column"
      />
      <div
        className="six wide column"
      >
        <h1>
          Currencies
        </h1>
        <p>
          click to select
        </p>
      </div>
      <div
        className="two wide column"
      />
      <div
        className="six wide column"
      >
        <h1>
          Selected Currencies
        </h1>
        <span>
          click to de-select
        </span>
        <button
          className="ui blue button deselect-button"
          disabled={undefined}
          onClick={[Function]}
          role="button"
          tabIndex={undefined}
        >
          deselect all
        </button>
      </div>
      <div
        className="one wide column"
      />
    </div>
    <div
      className="row"
    >
      <div
        className="one wide column"
      />
      <div
        className="six wide column"
      >
        <div
          className="ui segment"
        />
      </div>
      <div
        className="two wide column"
      />
      <div
        className="six wide column"
      >
        <div
          className="ui segment"
        />
      </div>
      <div
        className="one wide column"
      />
    </div>
  </div>
</div>
`;
